Make Player.HasVideo exclusive for video

garbear opened the pull request:
Currently, `Player.HasVideo` will return true if a game is being played. This causes a bug in the FullscreenGame window where the `Fullscreen` action attempts to open FullscreenVideo instead of leaving FullscreenGame for the GUI.

## Motivation and Context
Technically, games have video. I relied on this technicality in the original RetroPlayer PR to reduce the amount of code. Just imagine trying to rebase these hundreds of lines of changes spread throughout Kodi every month...

Thanks to @AchimTuran for reporting this bug during DevCon.

## How Has This Been Tested?
The bug causing the following screenshot no longer occurs on OSX.

The logic refactors will need another set of eyes. When I split FullscreenGame and FullscreenVideo, a small bug snuck in, so it's possible I missed something again.

## Screenshots (if appropriate):
<img width="1073" alt="screen shot 2017-11-02 at 11 30 51 pm" src="https://user-images.githubusercontent.com/531482/32361802-073c835a-c021-11e7-986a-6ee3a2521423.png">

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

garbear commented:
ping @Montellese for the JSON-RPC change. Do we want this?

FernetMenta commented on xbmc/ApplicationPlayer.cpp:
```diff
@@ -224,12 +224,12 @@ bool CApplicationPlayer::IsPausedPlayback()
 bool CApplicationPlayer::IsPlayingAudio() const
 {
-  return (IsPlaying() && !HasVideo() && HasAudio());
+  return (IsPlaying() && !HasVideo() && !HasGame() && HasAudio());
```
this is a perfect indicator for a weak design.

FernetMenta commented:
> Make Player.HasVideo exclusive for video

you changed 14 files for this intent. imo this change made the already weak design even worse.
instead of trying to make clients aware of player's state, we should have players (or ApplicationPlayer) control what should happen.

garbear commented:
How would this look?

FernetMenta commented:
let's take CApplication::CheckScreenSaverAndDPMS as an example. Currently this method has to be aware of all possible reasons that may inhibit screen saver or dpms. CApplicationPlayer knows player state. It can place an inhibitor for not going into DPMS without CApplication having to know the reason.

CApplication::OnAction
this entire condition checking should go away. players know when to handle a command. if a component handles the action, it signals back to app.

...

garbear commented:
Where's a good place to start for moving state checks into the player?

FernetMenta commented:
why don't you start with the two examples I gave?

garbear committed:
Clean up player notification logic in CApplication::OnAction()

garbear committed:
Improve logic in CAutorun::ExecuteAutorun()

garbear committed:
Improve logic for GUI_MSG_QUEUE_NEXT_ITEM in CApplication::OnMessage()

garbear committed:
Make Player.HasVideo exclusive for video

Currently, Player.HasVideo returns true for games. This causes a bug when
the Fullscreen action is executed during gameplay from the FullscreenGame
window.

garbear committed:
Improve sleep time logic in CApplication::FrameMove()

garbear committed:
Improve logic in CGraphicContext::SetFullScreenVideo()

garbear committed:
Fix game not shown in background when leaving fullscreen

garbear committed:
Python: Add isPlayingGame() function

garbear committed:
JSON-RPC: Add "Game" player type

garbear committed:
Improve logic in CApplication::CheckScreenSaverAndDPMS()

garbear committed:
Improve logic in CApplication::ActivateScreenSaver()
